SmartDeviceLink/SDLCacheFileManager.m

```diff
@@ -0,0 +1,282 @@
+                    iconArchiveFile = [[SDLIconArchiveFile alloc] init];
+                    break;
+                }
+                
+                return completion(icon, nil);
+            }
+        }
+    }
+    
+    // Download Icon
+    [self sdl_downloadIconFromRequestURL:request.url withCompletionHandler:^(UIImage * _Nullable image, NSError * _Nullable error) {
+        if (error != nil) {
+            return completion (nil, error);
+        }
+        
+        // Save lock screen icon to file path
+        NSString *iconFilePath = [self sdl_writeImage:image toFileFromURL:request.url];
+
+        if (iconFilePath == nil) {
+            SDLLogE(@"Could not save lockscreen icon to path: %@", error);
+        }
+        
+        // Update archive file with icon
+        [self updateArchiveFileFromRequest:request iconFilePath:iconFilePath archiveFile:iconArchiveFile];
+        
+        return completion(image, nil);
+    }];
+}
+
+#pragma mark - Cache Helper Methods
+
+/**
+ * Handles request to retrieve lock screen archive file from specified path.
+ *
+ * @param path The directory where the archive file is held
+ * @return The archive file at path
+*/
+- (nullable SDLIconArchiveFile *)sdl_retrieveArchiveFileFromPath:(NSString *)path {
+    NSString *archiveObjectPath = self.sdl_archiveFileDirectory;
+    SDLIconArchiveFile *iconArchiveFile = [NSKeyedUnarchiver unarchiveObjectWithFile:archiveObjectPath];
+    
+    return iconArchiveFile;
+}
+
+/**
+ * Handles the request to clear all files from lock screen icon directory in cache.
+*/
+- (void)sdl_clearLockScreenDirectory {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    NSString *directory = self.sdl_cacheFileBaseDirectory;
+    NSError *error = nil;
+    
+    for (NSString *file in [fileManager contentsOfDirectoryAtPath:directory error:&error]) {
+        if (error != nil) {
+            SDLLogE(@"Failed retrieving contents of directory: %@ with error: %@", directory, error);
+        }
+        BOOL success = [fileManager removeItemAtPath:[directory stringByAppendingPathComponent:file] error:&error];
+        if (!success || error != nil) {
+            SDLLogE(@"Could not delete file: %@", error);
+            break;
+        }
+    }
+}
+
+/**
+ * Handles request to save icon image to cache.
+ *
+ * @param icon The icon to be saved to the cache
+ * @param urlString The url of the icon to be hashed and used as file name
+ * @return A string representation of the directory the icon was saved to
+*/
+- (nullable NSString *)sdl_writeImage:(UIImage *)icon toFileFromURL:(NSString *)urlString {
+    NSData *iconPNGData = UIImagePNGRepresentation(icon);
+    NSString *iconStorageName = [self.class sdl_md5HashFromString:urlString];
+    
+    NSString *imageFilePath = [self.sdl_cacheFileBaseDirectory stringByAppendingPathComponent:iconStorageName];
+    [iconPNGData writeToFile:imageFilePath atomically:YES];
```

Comment:
If this returns `NO`, the operation failed, we should return an error and `nil` if it failed.

Reply:
I made this method's return type nullable to check for nil when used to handle any errors. Is it better practice to write a completion handler and return an error that way?

Reply:
Not necessarily. If it's a synchronous operation, then you should just return and not use a completion handler.

Reply:
Just return log an error and return nil then like below?
```   
BOOL writeSuccess = [iconPNGData writeToFile:imageFilePath atomically:YES];
if (!writeSuccess) {
        SDLLogE(@"Could not write file to specified path: %@", error);
        return nil;
}
    
return imageFilePath;

Reply:
Right